docarray/array/array.py

```diff
@@ -13,34 +32,299 @@ class DocumentArray(
     BaseNode,
 ):
     """
-    a DocumentArray is a list-like container of Document of the same schema
+    a DocumentArray is a container of Document of the same schema.
 
     :param docs: iterable of Document
+
+
+    A DocumentArray can only contain Documents that follow the same schema. To precise
+    this schema you can use the `DocumentArray[Document]` syntax. This will create a
+    DocumentArray that can only contain Document of the type Document. (Note that there
+    exists a special schema (AnySchema) that allows any Document to be stored in the
+    DocumentArray, but this is not recommended to use).
+
+    EXAMPLE USAGE
+    .. code-block:: python
+        from docarray import Document, DocumentArray
+        from docarray.typing import NdArray, ImageUrl
+
+
+        class Image(Document):
+            tensor: Optional[NdArray[100]]
+            url: ImageUrl
+
+
+        da = DocumentArray[Image](Image(url='http://url.com') for _ in range(10))
+
+
+    DocumentArray defines setter and getter for each field of the Document schema. These
+    getters and setters are defined dynamically at runtime. This allows to access the
+    field of the Document in a natural way. For example, if you have a DocumentArray of
+    Image you can do: `da.tensor` to get the tensor of all the Image in the
+    DocumentArray. You can also do `da.tensor = np.random.random([10, 100])` to set the
+    tensor of all the Image.
+
+
+    A DocumentArray can be in one of two modes: unstacked mode and stacked mode.
+
+
+    **Unstacked mode (default)**:
+    In this case a DocumentArray is a list of Document and each Document owns its data.
+    The getter and setter shown above will return a list of the fields of each Document
+    (or a DocumentArray if the field is a nested Document). This list/DocumentArray will
+    be created on the fly. The setter will set the field of each Document to the value
+    of the list/DocumentArray/Tensor passed as parameters.
+
+    Nevertheless, this list-like behavior is not always optimal especially when you want
+    to process you data in batch and do operation which involves matrix computation,This
+    is where the stack mode of the DocumentArray comes in handy.
+
+    In **stacked mode** each field which are Tensor are stored as a column in a tensor
+    of batch the size of the DocumentArray. This allows to do operation on the whole
+    batch instead of iterating over the DocumentArray.
+    In this mode the Document inside in the Document don't own the data anymore but just
+    reference to the data in the tensor of the DocumentArray.
+    In stacked mode the getter and setter just replace the tensor of the DocumentArray
+    of the given field.
```

Comment:
Again, even I don't know what these getters and setters are supposed to be honestly. Let's just not talk about it